README.md
[![npm version](https://badge.fury.io/js/%40amraneze%2Freact-instagram-login.svg)](https://badge.fury.io/js/%40amraneze%2Freact-instagram-login)

# React Instagram Login

> An Instagram oAUth Sign-in / Log-in Component for React

## Install

```
npm install react-instagram-login

```

## How to use

### With default button

```js
import React from "react";
import ReactDOM from "react-dom";
import InstagramLogin from "react-instagram-login";

const responseInstagram = (response) => {
  console.log(response);
};

ReactDOM.render(
  <InstagramLogin
    clientId="5fd2f11482844c5eba963747a5f34556"
    buttonText="Login"
    onSuccess={responseInstagram}
    onFailure={responseInstagram}
  />,
  document.getElementById("instagramButton")
);
```

### With a custom button

```js
import React from 'react';
import ReactDOM from 'react-dom';
import InstagramLogin from 'react-instagram-login';

const responseInstagram = (response) => {
  console.log(response);
}

ReactDOM.render(
  <InstagramLogin
    clientId="5fd2f11482844c5eba963747a5f34556"
    buttonText="Login"
    onSuccess={responseInstagram}
    onFailure={responseInstagram}
    render={({onClick} => (<button onClick={onClick}>Login with Instagram</button>))}
  />,
  document.getElementById('instagramButton')
);
```

## onSuccess callback

### Displaying OAuth using a popup

> Note: The redirectUri needs to be the same url as the current url.

### Displaying OAuth using a redirection

If you want to use redirection you should change the prop `useRedirect` to true.

Callback will return a code for use on your server to get a full access_token.

If `implicitAuth` is set to `true` it will return the full access_token directly.

## onFailure callback

Callback will return an error object.

|   property name   | value  |
| :---------------: | :----: |
|       error       | string |
|   error_reason    | string |
| error_description | string |

## Parameters

|    params    |  value   |    default value     |
| :----------: | :------: | :------------------: |
|   clientId   |  string  |       REQUIRED       |
|    scope     |  string  |        basic         |
|  onSuccess   | function |       REQUIRED       |
|  onFailure   | function |       REQUIRED       |
| redirectUri  |  string  |          -           |
|  buttonText  |  string  | Login with Instagram |
|   cssClass   |  string  |          -           |
|     tag      |  string  |        button        |
|     type     |  string  |        button        |
| implicitAuth | boolean  |        false         |
| useRedirect  | boolean  |        false         |
|    width     |  number  |         400          |
|    height    |  number  |         800          |
|    render    | function |          -           |

Instagram API Docs: https://www.instagram.com/developer/

You can now also pass child components such as icons into the button component.

```js
<InstagramLogin
  clientId="5fd2f11482844c5eba963747a5f34556"
  onSuccess={responseInstagram}
  onFailure={responseInstagram}
>
  <FontAwesome name="instagram" />
  <span> Login with Instagram</span>
</InstagramLogin>
```

## Dev Server

```
npm run start

```

## Run Tests

```
npm run test:watch

```

## Production Bundle

```
npm run bundle
```

#### TODO

- [ ] Use Hooks
- [ ] Fix the test cases
- [ ] Update dependencies
- [ ] Create a babel Conf
- [ ] Use prettier + Husky
- [ ] Improve and refactor the code
- [ ] Use Google, Facebook, Linkedin ... in this project

### Follow me on Twitter: [@Amraneze](https://twitter.com/amraneze)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.1] - 2020-12-08

### Added

- Add `this` to webpack output instead of `self`

## [2.0.0] - 2020-12-07

### Added

- Update README.md
- Change instgram button
- Add a popup to get the code with a success callback
- Add a render param to let the user to use their own custom button

[2.0.0]: https://github.com/Amraneze/react-instagram-login/releases/tag/2.0.0


src/index.js
import React, { Component } from 'react'
import PropTypes from 'prop-types' // eslint-disable-line import/no-extraneous-dependencies
import openPopup from './popup'

function getQueryVariable(context, variable) {
  const query = context.location.search.substring(1)
  const vars = query.split('&')
  const code = vars
    .map(i => {
      const pair = i.split('=')
      if (pair[0] === variable) {
        return pair[1]
      }

      return null
    })
    .filter(d => {
      if (d) {
        return true
      }

      return false
    })

  return code[0]
}

class InstagramLogin extends Component {
  constructor(props) {
    super(props)
    this.state = {
      hover: false
    }
    this.onBtnClick = this.onBtnClick.bind(this)
  }

  componentDidMount() {
    this.checkInstagramAuthentication(window)
  }

  onBtnClick() {
    const { clientId, scope } = this.props
    const redirectUri = this.props.redirectUri || window.location.href
    const responseType = this.props.implicitAuth ? 'token' : 'code'
    const url = `https://api.instagram.com/oauth/authorize?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=${responseType}&scope=${scope}`
    if (this.props.useRedirect) {
      window.location.href = url
    } else {
      this.oAuthSignIn({ url })
    }
  }

  // eslint-disable-next-line consistent-return
  onCredentialsChanged(popup, resolve, reject) {
    if (!resolve) {
      return new Promise((res, rej) => this.onCredentialsChanged(popup, res, rej))
    }
    let isFinished
    try {
      isFinished = this.checkInstagramAuthentication(popup)
    } catch (err) {
      // An exception is thrown when we try to access to another website's url
    }

    if (isFinished) {
      popup.close()
    } else if (popup.closed) {
      this.props.onFailure({
        error: 'closed',
        error_reason: 'oauth_canceled',
        error_description: 'User canceled the authentication'
      })
    } else {
      setTimeout(() => this.onCredentialsChanged(popup, resolve, reject), 0)
    }
  }

  oAuthSignIn({ url, tab = false }) {
    const name = tab ? '_blank' : 'instagram'
    const { width, height } = this.props
    const popup = openPopup({ url, name, width, height })
    this.onCredentialsChanged(popup)
  }

  checkInstagramAuthentication(context) {
    if (this.props.implicitAuth) {
      const matches = context.location.hash.match(/=(.*)/)
      if (matches) {
        this.props.onSuccess(matches[1])

        return true
      }
    } else if (context.location.search.includes('code')) {
      this.props.onSuccess(getQueryVariable(context, 'code'))

      return true
    } else if (context.location.search.includes('error')) {
      this.props.onFailure({
        error: getQueryVariable(context, 'error'),
        error_reason: getQueryVariable(context, 'error_reason'),
        error_description: getQueryVariable(context, 'error_description')
      })

      return true
    }

    return false
  }

  render() {
    const style = {
      display: 'block',
      border: 0,
      borderRadius: 3,
      boxShadow: 'rgba(0, 0, 0, 0.5) 0 1px 2px',
      color: '#ffffff',
      cursor: 'pointer',
      fontSize: '19px',
      overflow: 'hidden',
      padding: '10px',
      userSelect: 'none',
      background: 'linear-gradient(to right, rgb(236, 146, 35) 0%, rgb(177, 42, 160) 50%, rgb(105, 14, 224) 100%)',
      hover: {
        background: 'linear-gradient(to right, rgb(214, 146, 60) 0%, rgb(160, 11, 143) 50%, rgb(56, 10, 115) 100%)'
      }
    }
    const { hover } = this.state
    const { cssClass, buttonText, children, tag, type, render } = this.props
    if (render) {
      return render({ onClick: this.onBtnClick })
    }
    const instagramLoginButton = React.createElement(
      tag,
      {
        className: cssClass,
        onClick: this.onBtnClick,
        onMouseEnter: () => {
          this.setState({ hover: true })
        },
        onMouseLeave: () => {
          this.setState({ hover: false })
        },
        style: cssClass ? {} : { ...style, ...(hover ? style.hover : null) },
        type
      },
      children || buttonText
    )

    return instagramLoginButton
  }
}
InstagramLogin.defaultProps = {
  buttonText: 'Login with Instagram',
  scope: 'instagram_graph_user_media',
  tag: 'button',
  type: 'button',
  implicitAuth: false,
  useRedirect: false,
  width: 400,
  height: 800
}

InstagramLogin.propTypes = {
  onSuccess: PropTypes.func.isRequired,
  onFailure: PropTypes.func.isRequired,
  clientId: PropTypes.string.isRequired,
  buttonText: PropTypes.string,
  scope: PropTypes.string,
  cssClass: PropTypes.string,
  children: PropTypes.node,
  tag: PropTypes.string,
  redirectUri: PropTypes.string,
  type: PropTypes.string,
  implicitAuth: PropTypes.bool,
  useRedirect: PropTypes.bool,
  width: PropTypes.number,
  height: PropTypes.number,
  render: PropTypes.func
}

export default InstagramLogin
